source/common/upstream/upstream_impl.cc

```diff
@@ -695,7 +695,7 @@ void PriorityStateManager::updateClusterPrioritySet(
   HostVectorSharedPtr hosts(std::move(current_hosts));
   LocalityWeightsMap empty_locality_map;
   LocalityWeightsMap& locality_weights_map =
-      priority_state_.empty() ? empty_locality_map : priority_state_[priority].second;
+      priority_state_.size() > priority ? priority_state_[priority].second : empty_locality_map;
```

Comment:
How do we get into this situation? Should there be some test that covers this?

Reply:
I don't actually know enough about the PriorityStateManager to say, but there was a test that was invoking this undefined behavior and getting lucky, I guess, because it continued to pass. I only saw it failing when I added the assert in https://github.com/envoyproxy/envoy/pull/3856/commits/89f27cb352d8ae90fdb31f328e4c333298e4cd03. @dio, thoughts?

Reply:
Sorry about that. Thanks, @akonradi for catching this. The relevant test is when we clearing endpoints in this test case: https://github.com/envoyproxy/envoy/blob/01d2e16e973a76cc05f5e9e5c7a74f73498bc6c2/test/common/upstream/eds_test.cc#L552

Reply:
Should we keep the assertion?

Reply:
Yeah it seems like there should be some test that should fail w/o this fix either via assertion or something else?

Reply:
I did some digging. We caught this during the Google import because our `std::vector::operator[]` implementation does a bounds check, which is not in the spec. Unfortunately, there's no clang sanitizer we can use to do this here since the std::vector implementation is in code, and indexing out-of-bounds doesn't trigger an ASAN violation if the vector allocated extra space.

It looks like both libstdc++ and libc++ [have their own ways to enable bounds checks](https://stackoverflow.com/questions/16467380/bounds-checking-of-stdvector-and-other-containers-in-clang?rq=1). I tried enabling `_GLIBCXX_DEBUG` locally but it causes compile errors elsewhere that I don't want to fix in this PR. Moving forward, we should probably define both `_GLIBCXX_DEBUG` and `_LIBCPP_DEBUG` for ASAN builds, but that's going to be a future PR. Happy to open an issue if that sounds reasonable.

For now, I've added back in the ASSERT. It feels a little redundant now that the bug has been fixed but it sounds like that's preferred over doing nothing.

Reply:
OK that's fine. There is another issue already opened on using _GLIBXX_DEBUG: https://github.com/envoyproxy/envoy/issues/2556